ci/tasks.py

```diff
@@ -0,0 +1,259 @@
+#  see https://github.com/kubernetes/kubernetes/pull/96084
+@task(
+    parse_calitp_config,
+    help={
+        "app": GENERIC_HELP["app"],
+        "secret": "Optionally, specify a single secret to deploy",
+    },
+)
+def secrets(
+    c,
+    app=None,
+    secret=None,
+    hide=True,
+):
+    """
+    Deploy secret(s) by channel, and optionally app or secret name.
+    """
+    client = secretmanager.SecretManagerServiceClient()
+    found_secret = False
+    for release in get_releases(c, app=app):
+        for release_secret in release.secrets:
+            if not secret or secret == release_secret:
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    secret_path = Path(tmpdir) / Path(f"{release_secret}.yml")
+                    # this ID maps to cal-itp-data-infra; there's probably a better way to do this
+                    # TODO: we could probably just use gcloud CLI for this
+                    name = f"projects/1005246706141/secrets/{release_secret}/versions/latest"
+                    secret_contents = client.access_secret_version(
+                        request={"name": name}
+                    ).payload.data.decode("UTF-8")
+
+                    if release.namespace:
+                        ns_str = f"--namespace {release.namespace}"
+                    else:
+                        ns_str = ""
+                        assert (
+                            "namespace" in yaml.safe_load(secret_contents)["metadata"]
+                        )
+
+                    with open(secret_path, "w") as f:
+                        f.write(secret_contents)
+                    print(f"Applying {release_secret}...", flush=True)
+                    result = c.run(
+                        f"kubectl apply {ns_str} -f {secret_path}", hide=hide, warn=True
+                    )
+                    if result.exited:
+                        print(
+                            f"FAILURE: Failed to apply secret {release_secret}.",
+                            flush=True,
+                        )
+                        raise Exit
+                    print(f"Successfully applied {release_secret}.", flush=True)
+                found_secret = True
+
+    if not found_secret:
+        print("WARNING: Failed to deploy any secrets.", flush=True)
+
+
+@task(
+    parse_calitp_config,
+    help={
+        **GENERIC_HELP,
+        "outfile": "File in which to save the combined kubectl diff output",
+    },
+)
+def diff(
+    c,
+    driver=None,
+    app=None,
+    outfile=None,
+):
+    """
+    Applies kubectl diff to manifests for a given channel.
+    """
+    assert c.calitp_config.git_root is not None
+    actual_driver = ReleaseDriver[driver] if driver else None
+    full_diff = ""
+    result: Result
+
+    for release in get_releases(c, driver=actual_driver, app=app):
+        if release.driver == ReleaseDriver.kustomize:
+            assert release.kustomize_dir is not None
```

Comment:
similarly, something here seems a bit weird re: having asserts on things that you defined above... should these attributes just be required in the relevant pydantic classes?

Reply:
An alternative would be to use a [discriminated union](https://docs.pydantic.dev/dev-v2/usage/types/unions/#discriminated-unions-aka-tagged-unions) but I think it's overkill here; most of these asserts are primarily to make mypy happy. I think if we were to implemented a third driver (e.g. just vanilla `kubectl`) then it'd be worth creating a union of 3 types.